GVFS/GVFS/CommandLine/StatisticsVerb.cs

```diff
@@ -0,0 +1,209 @@
+            List<string> topLevelDirectoriesByHydration = enlistmentStatistics.GetDirectoriesSortedByHydration();
+
+            this.Output.WriteLine("\nRepository statistics");
+            this.Output.WriteLine("Total paths tracked by git:     " + trackedFilesCountFormatted.PadLeft(longest) + " | 100%");
+            this.Output.WriteLine("Total number of placeholders:   " + placeholderCountFormatted.PadLeft(longest) + " | " + this.FormatPercent(enlistmentStatistics.GetPlaceholderPercentage()));
+            this.Output.WriteLine("Total number of modified paths: " + modifiedPathsCountFormatted.PadLeft(longest) + " | " + this.FormatPercent(enlistmentStatistics.GetModifiedPathsPercentage()));
+
+            this.Output.WriteLine("\nTotal hydration percentage:     " + this.FormatPercent(enlistmentStatistics.GetPlaceholderPercentage() + enlistmentStatistics.GetModifiedPathsPercentage()).PadLeft(longest + 7));
+
+            this.Output.WriteLine("\nMost hydrated top level directories:");
+
+            int maxDirectoryNameLength = 0;
+            for (int i = 0; i < 5 && i < topLevelDirectoriesByHydration.Count; i++)
+            {
+                maxDirectoryNameLength = Math.Max(maxDirectoryNameLength, topLevelDirectoriesByHydration[i].Length);
+            }
+
+            for (int i = 0; i < DirectoryDisplayCount && i < topLevelDirectoriesByHydration.Count; i++)
+            {
+                string dir = topLevelDirectoriesByHydration[i].PadRight(maxDirectoryNameLength);
+                string percent = this.FormatPercent(enlistmentStatistics.GetHydrationOfDirectory(topLevelDirectoriesByHydration[i]));
+                this.Output.WriteLine(" " + percent + " | " + dir);
+            }
+
+            bool healthyRepo = (enlistmentStatistics.GetPlaceholderPercentage() + enlistmentStatistics.GetModifiedPathsPercentage()) < (0.5);
+
+            this.Output.WriteLine("\nRepository status: " + (healthyRepo ? "Healthy" : "Unhealthy"));
+        }
+
+        /// <summary>
+        /// Takes a fractional double and formats it as a percent taking exactly 4 characters with no decimals
+        /// </summary>
+        /// <param name="percent">Fractional double to format to a percent</param>
+        /// <returns>A 4 character string formatting the percent correctly</returns>
+        private string FormatPercent(double percent)
+        {
+            return percent.ToString("P0").PadLeft(4);
+        }
+
+        /// <summary>
+        /// Parse a line of the git index coming from the ls-tree endpoint in the git process to get the path to that file
+        /// </summary>
+        /// <param name="line">The line from the output of the git index</param>
+        /// <returns>The path extracted from the provided line of the git index</returns>
+        private string TrimGitIndexLine(string line)
+        {
+            return line.Substring(line.IndexOf("blob") + 46);

```

Comment:
Why do we add 46?

Reply:
In order to make the GVFSEnlistmentStatistics class the most accessible and flexible, I wanted it to just take `List<string>` objects with all of the paths. If it takes `IPlaceholderData` then it only works for this one particular instance of the system, where I feel that `List<string>` is more likely to be accessible. If this were python, I'd use the `map()` function to modify it in place. Is there something similar for c#?

Reply:
46 is the length of 'blob' + some whitespace + the length of the git hash + some more whitespace to take the cursor to the start of the path (at the end of the line spit out by the index). Maybe there is a way to just consume everything from the end up to whitespace? But then you lose support for whitespace in file names

Reply:
I think we can use the documentation for the format and use that to parse instead of some magic number.  See [Output Format](https://git-scm.com/docs/git-ls-tree)
```
<mode> SP <type> SP <object> TAB <file>
```

Reply:
That's a good idea. I replaced it with `line.Substring(line.IndexOf('\t') + 1);` and that worked (and looks much better)